leptonai/cli/storage.py

```diff
@@ -297,25 +317,47 @@ def upload(local_path, remote_path, rsync, recursive, progress, file_system):
             flags += "a"
         if progress:
             flags += " --progress"
-        command = (
-            f"rsync {flags}"
-            f" {local_path} rsync://{workspace_id}@{ip}:{port}/volume{remote_path}"
-        )
-        console.print(f"Running command: [bold]{command}[/]")
-
-        process = subprocess.Popen(
-            command,
-            stdout=subprocess.PIPE,
-            stderr=subprocess.STDOUT,
-            env=env_vars,
-            shell=True,
-            universal_newlines=True,
-        )
 
-        for line in process.stdout:
-            print(line, end="")
-        process.wait()
+        attempts = auto_recover
+        while attempts > 0:
+            cur_flags = flags + " --partial" if attempts > 1 else flags
+            command = (
+                f"rsync {cur_flags}"
+                f" {local_path} rsync://{workspace_id}@{ip}:{port}/volume{remote_path}"
+            )
+            console.print(f"Running command: [bold]{command}[/]")
+
+            process = subprocess.Popen(
+                command,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT,
+                env=env_vars,
+                shell=True,
+                universal_newlines=True,
+            )
+
+            for line in process.stdout:
+                print(line, end="")
+            process.wait()
+
+            return_code = process.returncode
+            if return_code != 0 and attempts > 0:
+                console.print(f"[red]Rsync failed[/] with exit code {return_code}.")
+                attempts -= 1
+                if attempts > 0:
+                    console.print("[green]Retrying... [/]")
+                    time.sleep(3)
+            else:
+                break
```

Comment:
seems the attempts counter doesn't match line 323.

based on line 323, `attempts == 1` is the last run, however, here we return only when `attempts == 0`